File: 1365/main.cpp
class Solution {
public:
    vector<int> smallerNumbersThanCurrent(vector<int>& nums) {
        vector<int> counts(101);
        vector<int> res;
        for(int& n : nums) counts[n]++;
        for(int i=1; i<=100; ++i) {
            counts[i] += counts[i-1];
        }

        for(int& n : nums) {
            if(n == 0) {
                res.push_back(0);
            } else {
                res.push_back(counts[n-1]);
            }
        }

        return res;
    }
};

File: 1365/main.rs
impl Solution {
    pub fn smaller_numbers_than_current(nums: Vec<i32>) -> Vec<i32> {
        let mut counts = vec![0i32; 101];

        for n in nums.iter() {
            counts[*n as usize] += 1;
        }

        for i in 1..101 {
            counts[i] += counts[i-1];
        }
        
        let mut res: Vec<i32> = Vec::new();
        for n in nums.iter() {
            match n {
                0 => res.push(0),
                _ => res.push(counts[(*n as usize) - 1]),
            }
        }

        res

       // let mut counts = vec![0; 102]; // 0
       // nums.iter().for_each(|&n| counts[1 + n as usize] += 1);
       // (1..counts.len()).for_each(|n| counts[n] += counts[n-1]);
       // nums.iter().map(|&n| counts[n as usize]).collect() // map is lazy evaluation
    }
}